website/src/pages/training/troubleshooting-ctf/scenario-2.md

```diff
@@ -0,0 +1,277 @@
+    Port:           <none>
+    Host Port:      <none>
+    State:          Running
+      Started:      Tue, 21 Feb 2023 21:49:43 +0000
+    Ready:          True
+    Restart Count:  0
+
+---
+  frontend:
+    Container ID:   containerd://f170cb91396dd1cd5161a2e9badd0d0831bee0df4cd5a9a9a96a4140ffa246a5
+    Image:          registry.gitlab.com/lionelj/meridio/frontend:v1.0.0
+    Image ID:       registry.gitlab.com/lionelj/meridio/frontend@sha256:520a12c4c3982dd0fe366f023cf86baf633cc8c7c56ae4f64f415a279efede12
+    Port:           <none>
+    Host Port:      <none>
+    State:          Running
+      Started:      Tue, 21 Feb 2023 21:49:52 +0000
+    Ready:          False
+    Restart Count:  0
+    Liveness:       exec [/bin/grpc_health_probe -addr=unix:///tmp/health.sock -service= -connect-timeout=250ms -rpc-timeout=350ms] delay=0s timeout=3s period=10s #success=1 #failure=5
+    Readiness:      exec [/bin/grpc_health_probe -addr=unix:///tmp/health.sock -service=Readiness -connect-timeout=250ms -rpc-timeout=350ms] delay=0s timeout=3s period=10s #success=1 #failure=5
+    Startup:        exec [/bin/grpc_health_probe -addr=unix:///tmp/health.sock -service= -connect-timeout=250ms -rpc-timeout=350ms] delay=0s timeout=2s period=2s #success=1 #failure=30
+```
+
+```bash
+  Warning  Unhealthy  62s (x4 over 64s)  kubelet            Readiness probe failed: service unhealthy (responded with "NOT_SERVING")
+  Warning  Unhealthy  53s (x5 over 64s)  kubelet            Readiness probe failed: service unhealthy (responded with "NOT_SERVING")
+```
+
+4. From the previous step, it can be concluded that the reason for the failed *readiness* of the two containers, `frontend` and `stateless-lb`, is the failed ***readiness probe***. To debug further, it would be useful to check the logs of the two containers for any sort of errors.
+   * Going through the logs of the `stateless-lb` container, it can be noticed that there are no ***error*** messages.
+   * Going through the logs of the `frontend` container, it can be noticed that there is one ***error*** message, which states `"error":"gateway down"`.
+
+
+```bash
+kubectl logs stateless-lb-frontend-attractor-a-1-d8db96c8f-q92dc -n=red -c=stateless-lb | grep "\"severity\":\"error\""
+
+...
+
+kubectl logs stateless-lb-frontend-attractor-a-1-d8db96c8f-q92dc -n=red -c=frontend | grep "\"severity\":\"error\""
+{"severity":"error","timestamp":"2023-02-21T21:49:56.733+00:00","service_id":"Meridio-frontend","message":"connectivity","version":"1.0.0","extra_data":{"class":"FrontEndService","func":"Monitor","status":16,"out":["BIRD 2.0.8 ready.","Name       Proto      Table      State  Since         Info","NBR-gateway-v4-a BGP        ---        start  21:49:53.760  Active            Neighbor address: 169.254.100.150%ext-vlan0","NBR-gateway-v6-a BGP        ---        start  21:49:53.760  Active            Neighbor address: 100:100::150%ext-vlan0",""],"error":"gateway down"}}
+```
+
```

Comment:
It is also possible to check the BGP status with some commands:
```
ubuntu $ kubectl exec -it stateless-lb-frontend-attractor-a-1-d8db96c8f-nsw9t -n red -c frontend -- birdc -s /var/run/bird/bird.ctl show protocol all
BIRD 2.0.8 ready.
Name       Proto      Table      State  Since         Info
device1    Device     ---        up     10:06:22.017  

kernel1    Kernel     master4    up     10:06:22.017  
  Channel ipv4
    State:          UP
    Table:          master4
    Preference:     10
    Input filter:   REJECT
    Output filter:  default_rt
    Routes:         0 imported, 0 exported, 0 preferred
    Route change stats:     received   rejected   filtered    ignored   accepted
      Import updates:              0          0          0          0          0
      Import withdraws:            0          0        ---          0          0
      Export updates:              0          0          0        ---          0
      Export withdraws:            0        ---        ---        ---          0

kernel2    Kernel     master6    up     10:06:22.017  
  Channel ipv6
    State:          UP
    Table:          master6
    Preference:     10
    Input filter:   REJECT
    Output filter:  default_rt
    Routes:         0 imported, 0 exported, 0 preferred
    Route change stats:     received   rejected   filtered    ignored   accepted
      Import updates:              0          0          0          0          0
      Import withdraws:            0          0        ---          0          0
      Export updates:              0          0          0        ---          0
      Export withdraws:            0        ---        ---        ---          0

NBR-gateway-v4-a BGP        ---        start  10:06:22.326  Active        Socket: Connection refused
  BGP state:          Active
    Neighbor address: 169.254.100.150%ext-vlan0
    Neighbor AS:      4248829953
    Local AS:         8103
    Connect delay:    2.089/5
    Last error:       Socket: Connection refused
  Channel ipv4
    State:          DOWN
    Table:          master4
    Preference:     100
    Input filter:   default_rt
    Output filter:  cluster_e_static
  Channel ipv6
    State:          DOWN
    Table:          master6
    Preference:     100
    Input filter:   REJECT
    Output filter:  REJECT

NBR-gateway-v6-a BGP        ---        start  10:06:22.326  Active        Socket: Connection refused
  BGP state:          Active
    Neighbor address: 100:100::150%ext-vlan0
    Neighbor AS:      4248829953
    Local AS:         8103
    Connect delay:    0.463/5
    Last error:       Socket: Connection refused
  Channel ipv4
    State:          DOWN
    Table:          master4
    Preference:     100
    Input filter:   REJECT
    Output filter:  REJECT
  Channel ipv6
    State:          DOWN
    Table:          master6
    Preference:     100
    Input filter:   default_rt
    Output filter:  cluster_e_static

NBR-BFD    BFD        ---        up     10:06:22.017
```
BGP state should be "Established", and States should be "UP". "Last error" can also be used.

Bird logs can also be used:
```
ubuntu $ kubectl exec -it stateless-lb-frontend-attractor-a-1-d8db96c8f-nsw9t -n red -c frontend -- cat /var/log/bird.log                            
2023-02-22 10:24:45.650 <TRACE> NBR-gateway-v4-a: Connecting to 169.254.100.150 from local address 169.254.100.2
2023-02-22 10:24:45.652 <TRACE> NBR-gateway-v4-a: Connection lost (Connection refused)
2023-02-22 10:24:45.652 <TRACE> NBR-gateway-v4-a: Connect delayed by 5 seconds
2023-02-22 10:24:46.141 <TRACE> NBR-gateway-v6-a: Connecting to 100:100::150 from local address 100:100::2
2023-02-22 10:24:46.141 <TRACE> NBR-gateway-v6-a: Connection lost (Connection refused)
2023-02-22 10:24:46.141 <TRACE> NBR-gateway-v6-a: Connect delayed by 5 second
```